ipld/unixfs/io/directory.go

```diff
@@ -104,30 +170,107 @@ type HAMTDirectory struct {
 	shard *hamt.Shard
 	dserv ipld.DAGService
 
+	// opts
+	maxLinks   int
+	cidBuilder cid.Builder
+	mode       os.FileMode
+	mtime      time.Time
+
 	// Track the changes in size by the AddChild and RemoveChild calls
 	// for the HAMTShardingSize option.
 	sizeChange int
+	totalLinks int
 }
 
-func newEmptyBasicDirectory(dserv ipld.DAGService) *BasicDirectory {
-	return newBasicDirectoryFromNode(dserv, format.EmptyDirNode())
+// NewBasicDirectory creates an empty basic directory with the given options.
+func NewBasicDirectory(dserv ipld.DAGService, opts ...DirectoryOption) *BasicDirectory {
+	basicDir := new(BasicDirectory)
+	basicDir.dserv = dserv
+
+	for _, o := range opts {
+		o(basicDir)
+	}
+
+	var node *mdag.ProtoNode
+	if basicDir.mode > 0 {
+		node = format.EmptyDirNodeWithStat(basicDir.mode, basicDir.mtime)
+	} else {
+		node = format.EmptyDirNode()
+	}
+	basicDir.node = node
+	basicDir.node.SetCidBuilder(basicDir.cidBuilder)
+
+	// Scan node links (if any) to restore estimated size.
+	basicDir.computeEstimatedSizeAndTotalLinks()
+
+	return basicDir
 }
 
-func newBasicDirectoryFromNode(dserv ipld.DAGService, node *mdag.ProtoNode) *BasicDirectory {
+// NewBasicDirectoryFromNode creates a basic directory wrapping the given
+// ProtoNode.
+func NewBasicDirectoryFromNode(dserv ipld.DAGService, node *mdag.ProtoNode) *BasicDirectory {
 	basicDir := new(BasicDirectory)
 	basicDir.node = node
 	basicDir.dserv = dserv
 
 	// Scan node links (if any) to restore estimated size.
-	basicDir.computeEstimatedSize()
+	basicDir.computeEstimatedSizeAndTotalLinks()
 
 	return basicDir
 }
 
-// NewDirectory returns a Directory implemented by DynamicDirectory
-// containing a BasicDirectory that can be converted to a HAMTDirectory.
-func NewDirectory(dserv ipld.DAGService) Directory {
-	return &DynamicDirectory{newEmptyBasicDirectory(dserv)}
+// NewHAMTDirectory creates an empty HAMT directory with the given options.
+func NewHAMTDirectory(dserv ipld.DAGService, sizeChange int, opts ...DirectoryOption) (*HAMTDirectory, error) {
+	dir := new(HAMTDirectory)
+	dir.dserv = dserv
+	dir.sizeChange = sizeChange
+	dir.maxLinks = 0
```

Comment:
`dir.maxLinks` should already be initialized to 0 when `dir` was created with `new`.
```suggestion
	dir := &HAMTDirectory{
		dserv: dserv,
		sizeChange: sizeChange,
	}
```